include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.h

```diff
@@ -68,19 +68,13 @@ struct PortInfo {
 
   /// Return true if this is an inout port.  This will be true if the port
   /// contains either bi-directional signals or analog types.
+  /// Non-HW types (e.g., ref types) are never considered InOut.
   bool isInOut() {
-    auto flags = TypeSwitch<Type, RecursiveTypeProperties>(type)
-                     .Case<FIRRTLBaseType>([](auto base) {
-                       return base.getRecursiveTypeProperties();
-                     })
-                     .Case<RefType>([](auto ref) {
-                       return ref.getType().getRecursiveTypeProperties();
-                     })
-                     .Default([](auto) {
-                       llvm_unreachable("unsupported type");
-                       return RecursiveTypeProperties{};
-                     });
-    return !flags.isPassive || flags.containsAnalog;
+    auto baseType = dyn_cast<FIRRTLBaseType>(type);
+    if (!baseType)
+      return false;
+
+    return !baseType.isPassive() || baseType.containsAnalog();
```

Comment:
Limiting "isInOut" to FIRRTLBaseType (~= "hardware type") works presently but conceivably other non-hardware-only types might want different answers.

For example, a non-hw aggregate type mixing hardware and not, this (and "is passive") is less clear, to be tackled when adding them (but heads-up).  Conservatively saying "no" for non-hw types (as we currently do).